File: src/dorkbox/network/dns/resolver/DnsNameResolverListResolverContext.java
/*
 * Copyright 2014 The Netty Project
 *
 * The Netty Project licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */
package dorkbox.network.dns.resolver;

import java.net.InetAddress;
import java.util.ArrayList;
import java.util.List;

import dorkbox.network.dns.resolver.addressProvider.DnsServerAddressStream;
import dorkbox.network.dns.resolver.cache.DnsCache;
import dorkbox.network.dns.resolver.cache.DnsCacheEntry;
import io.netty.util.concurrent.Promise;

/**
 *
 */
final
class DnsNameResolverListResolverContext extends DnsNameResolverContext<List<InetAddress>> {
    DnsNameResolverListResolverContext(DnsNameResolver parent,
                                       String hostname,
                                       DnsCache resolveCache,
                                       DnsServerAddressStream nameServerAddrs) {
        super(parent, hostname, resolveCache, nameServerAddrs);
    }

    @Override
    DnsNameResolverContext<List<InetAddress>> newResolverContext(DnsNameResolver parent,
                                                                 String hostname,
                                                                 DnsCache resolveCache,
                                                                 DnsServerAddressStream nameServerAddrs) {
        return new DnsNameResolverListResolverContext(parent, hostname, resolveCache, nameServerAddrs);
    }

    @Override
    boolean finishResolve(Class<? extends InetAddress> addressType,
                          List<DnsCacheEntry> resolvedEntries,
                          Promise<List<InetAddress>> promise) {

        List<InetAddress> result = null;
        final int numEntries = resolvedEntries.size();
        for (int i = 0; i < numEntries; i++) {
            final InetAddress a = resolvedEntries.get(i).address();
            if (addressType.isInstance(a)) {
                if (result == null) {
                    result = new ArrayList<InetAddress>(numEntries);
                }
                result.add(a);
            }
        }

        if (result != null) {
            promise.trySuccess(result);
            return true;
        }
        return false;
    }
}

File: src/dorkbox/dns/dns/resolver/DnsNameResolverListResolverContext.kt
/*
 * Copyright 2021 dorkbox, llc
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package dorkbox.dns.dns.resolver

import dorkbox.dns.dns.resolver.addressProvider.DnsServerAddressStream
import dorkbox.dns.dns.resolver.cache.DnsCache
import dorkbox.dns.dns.resolver.cache.DnsCacheEntry
import io.netty.util.concurrent.Promise
import java.net.InetAddress

/**
 *
 */
internal class DnsNameResolverListResolverContext(parent: DnsNameResolver, hostname: String, resolveCache: DnsCache, nameServerAddrs: DnsServerAddressStream) : DnsNameResolverContext<List<InetAddress>>(
    parent, hostname, resolveCache, nameServerAddrs
) {
    override fun newResolverContext(parent: DnsNameResolver, hostname: String, resolveCache: DnsCache, nameServerAddrs: DnsServerAddressStream): DnsNameResolverContext<List<InetAddress>> {
        return DnsNameResolverListResolverContext(parent, hostname, resolveCache, nameServerAddrs)
    }

    override fun finishResolve(addressType: Class<out InetAddress>, resolvedEntries: List<DnsCacheEntry>, promise: Promise<List<InetAddress>>): Boolean {
        var result: MutableList<InetAddress>? = null
        val numEntries = resolvedEntries.size

        for (i in 0 until numEntries) {
            val a = resolvedEntries[i].address()
            if (addressType.isInstance(a)) {
                if (result == null) {
                    result = ArrayList(numEntries)
                }
                result.add(a!!)
            }
        }
        if (result != null) {
            promise.trySuccess(result)
            return true
        }
        return false
    }
}
